operator/builtin/parser/csv/csv_test.go

```diff
@@ -0,0 +1,304 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package csv
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/open-telemetry/opentelemetry-log-collection/entry"
+	"github.com/open-telemetry/opentelemetry-log-collection/operator"
+	"github.com/open-telemetry/opentelemetry-log-collection/testutil"
+)
+
+var testHeader = "name,sev,msg"
+
+func newTestParser(t *testing.T) *CSVParser {
+	cfg := NewCSVParserConfig("test")
+	cfg.Header = testHeader
+	ops, err := cfg.Build(testutil.NewBuildContext(t))
+	require.NoError(t, err)
+	op := ops[0]
+	return op.(*CSVParser)
+}
+
+func TestCSVParserBuildFailure(t *testing.T) {
+	cfg := NewCSVParserConfig("test")
+	cfg.OnError = "invalid_on_error"
+	_, err := cfg.Build(testutil.NewBuildContext(t))
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "invalid `on_error` field")
+}
+
+func TestCSVParserBuildFailureInvalidDelimiter(t *testing.T) {
+	cfg := NewCSVParserConfig("test")
+	cfg.Header = testHeader
+	cfg.FieldDelimiter = ";;"
+	_, err := cfg.Build(testutil.NewBuildContext(t))
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "Invalid 'delimiter': ';;'")
+}
+
+func TestCSVParserStringFailure(t *testing.T) {
+	parser := newTestParser(t)
+	_, err := parser.parse("invalid")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "record on line 1: wrong number of fields")
+}
+
+func TestCSVParserByteFailure(t *testing.T) {
+	parser := newTestParser(t)
+	_, err := parser.parse([]byte("invalid"))
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "record on line 1: wrong number of fields")
+}
+
+func TestCSVParserInvalidType(t *testing.T) {
+	parser := newTestParser(t)
+	_, err := parser.parse([]int{})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "type '[]int' cannot be parsed as csv")
+}
+
+func TestParserCSV(t *testing.T) {
+	cases := []struct {
+		name       string
+		configure  func(*CSVParserConfig)
+		inputBody  interface{}
+		outputBody interface{}
+	}{
+		{
+			"basic",
+			func(p *CSVParserConfig) {
+				p.Header = testHeader
+			},
+			"stanza,INFO,started agent",
+			map[string]interface{}{
+				"name": "stanza",
+				"sev":  "INFO",
+				"msg":  "started agent",
+			},
+		},
+		{
+			"advanced",
+			func(p *CSVParserConfig) {
+				p.Header = "name;address;age;phone;position"
+				p.FieldDelimiter = ";"
+			},
+			"stanza;Evergreen;1;555-5555;agent",
+			map[string]interface{}{
+				"name":     "stanza",
+				"address":  "Evergreen",
+				"age":      "1",
+				"phone":    "555-5555",
+				"position": "agent",
+			},
+		},
+		{
+			"mariadb-audit-log",
+			func(p *CSVParserConfig) {
+				p.Header = "timestamp,serverhost,username,host,connectionid,queryid,operation,database,object,retcode"
+			},
+			"20210316 17:08:01,oiq-int-mysql,load,oiq-int-mysql.bluemedora.localnet,5,0,DISCONNECT,,,0",
+			map[string]interface{}{
+				"timestamp":    "20210316 17:08:01",
```

Comment:
I'm not sure what you mean.

Reply:
When a user configures the `csv_parser`, they might include a `timestamp` block:
```yaml
- type: csv_parser
  header: 'timestamp_field,severity,message'
  timestamp:
    parse_from: timestamp_field
    layout_type: strptime
    layout: '%Y-%m-%d'
```
Since we have a timestamp field here, we might as well demonstrate this functionality in the tests.